src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -367,135 +376,383 @@ public class SqlMutationEngine : IMutationEngine
-                            subStatusCode: DataApiBuilderException.SubStatusCodes.DatabasePolicyFailure
-                            );
+                        throw _dabExceptionWithTransactionErrorMessage;
                     }
 
-                    string primaryKeyRoute = ConstructPrimaryKeyRoute(context, mutationResultRow.Columns, sqlMetadataProvider);
-                    // location will be updated in rest controller where httpcontext is available
-                    return new CreatedResult(location: primaryKeyRoute, OkMutationResponse(mutationResultRow.Columns).Value);
-                }
+                    // When read permission is configured without a database policy, a subsequent select query will not be executed.
+                    // So, if the read action has include/exclude fields configured, additional fields present in the response
+                    // need to be removed.
+                    if (isReadPermissionConfiguredForRole && !isDatabasePolicyDefinedForReadAction)
+                    {
+                        IEnumerable<string> allowedExposedColumns = _authorizationResolver.GetAllowedExposedColumns(context.EntityName, roleName, EntityActionOperation.Read);
+                        foreach (string columnInResponse in mutationResultRow!.Columns.Keys)
+                        {
+                            if (!allowedExposedColumns.Contains(columnInResponse))
+                            {
+                                mutationResultRow!.Columns.Remove(columnInResponse);
+                            }
+                        }
+                    }
 
-                if (context.OperationType is EntityActionOperation.Update || context.OperationType is EntityActionOperation.UpdateIncremental)
-                {
-                    // Nothing to update means we throw Exception
-                    if (mutationResultRow is null || mutationResultRow.Columns.Count == 0)
+                    string primaryKeyRouteForLocationHeader = isReadPermissionConfiguredForRole ? ConstructPrimaryKeyRoute(context, mutationResultRow!.Columns, sqlMetadataProvider)
+                                                                                                : string.Empty;
+
+                    if (context.OperationType is EntityActionOperation.Insert)
                     {
-                        throw new DataApiBuilderException(message: "No Update could be performed, record not found",
-                                                           statusCode: HttpStatusCode.PreconditionFailed,
-                                                           subStatusCode: DataApiBuilderException.SubStatusCodes.DatabaseOperationFailed);
+                        // Location Header is made up of the Base URL of the request and the primary key of the item created.
+                        // However, for PATCH and PUT requests, the primary key would be present in the request URL. For POST request, however, the primary key
```

Comment:
This section does not deal with PATCH and PUT requests, so, doesnt make sense to mention those here to avoid confusion